source/postprocess/particle_density_histogram.cc

```diff
@@ -0,0 +1,268 @@
+        auto last_particle_in_cell = this->get_particle_manager(particle_manager_index).get_particle_handler().particles_in_cell(cell).end();
+        
+        for(auto particle_iterator=first_particle_in_cell; particle_iterator != last_particle_in_cell; std::advance(particle_iterator,1))
+        {
+          double particle_x = particle_iterator->get_reference_location()[0];
+          double particle_y = particle_iterator->get_reference_location()[1];
+
+          double x_ratio = (particle_x) / (bucket_width);
+          double y_ratio = (particle_y) / (bucket_width);
+
+          unsigned int x_index = static_cast<unsigned int>(std::floor(x_ratio));
+          unsigned int y_index = static_cast<unsigned int>(std::floor(y_ratio));
+
+          /*
+          if a particle is exactly on the boundary of two cells it's 
+          reference location will equal 1 and if this is the case
+          the x/y/z_index will be outside of the range of the table without
+          these checks.
+          */
+          if (x_index == (dim-1)){
+            x_index = dim-1;
+          }
+          if (y_index == (dim-1)){
+            y_index = dim-1;
+          }
+
+          TableIndices<dim> entry_index;
+          entry_index[0] = x_index;
+          entry_index[1] = y_index;
+          if (dim == 3){
+            const double particle_z = particle_iterator->get_reference_location()[2];
+            double z_ratio = (particle_z) / (bucket_width);
+            unsigned int z_index = static_cast<unsigned int>(std::floor(z_ratio));
+            if (z_index == (dim-1)){
+              z_index = dim-1;
+            }
+            entry_index[2] = z_index;
+          }
+
+          unsigned int particles_in_bucket = buckets(entry_index);
+          particles_in_bucket++;
+          buckets(entry_index) = particles_in_bucket;    
+        }       
+      }
+    }
+  
+  
+    template <int dim>
+    void
+    ParticleDensityStatistics<dim>::declare_parameters (ParameterHandler &prm)
+    {
+      prm.enter_subsection("Postprocess");
+      {
+        prm.enter_subsection("Particle Density Statistics");
+        {
+          prm.declare_entry("Histogram Granularity","2",
+                            Patterns::Integer (0),
+                            "This parameter determines how many bins the histogram plugin sorts particles "
+                            "into. The ideal value for granularity depends on the maximum number of particles "
+                            "in the cell. Generally higher values lead to more accuracy, but excessively high "
+                            "granularity can mean there are too few particles in each bin."
+                            );
+        }
+        prm.leave_subsection ();
+      }
+      prm.leave_subsection ();
+    }
+    
+    template <int dim>
+    void
+    ParticleDensityStatistics<dim>::parse_parameters (ParameterHandler &prm)
+    {
+      CitationInfo::add("particle density statistics");
```

Comment:
this line is unnecessary until you have written your thesis, and then you need to add several other things, before it does something ;-) ... leave it out for now.